source/Preprocessor.cpp

```diff
@@ -160,13 +161,72 @@ Token Preprocessor::nextRaw(LexerMode mode) {
 }
 
 Trivia Preprocessor::handleIncludeDirective(Token directive) {
-    // next token should be a filename
-    // TODO: handle macro replaced include file name
+    // next token should be a filename, or something the preprocessor generated
+    // that we will make into a filename
     Token fileName = next(LexerMode::IncludeFileName);
-    Token end = parseEndOfDirective();
+
+    // If we have a macro-expanded filename, there won't be an EOD token
+    bool suppressEODError = false;
+    if (fileName.kind != TokenKind::IncludeFileName) {
+        suppressEODError = true;
+        // A raw include file name will always yield a IncludeFileName token.
+        // a (valid) macro-expanded include filename will be lexed as either
+        // a StringLiteral or the token sequence '<' ... '>'
+        bool angleBrackets = false;
+        if (fileName.kind == TokenKind::LessThan) {
+            angleBrackets = true;
+            // In this case, we know that the first token is LessThan, and
+            // if things are proper, the last token is >, but there can be an
+            // arbitrary number of tokens that the macro expanded to in-between
+            // (as file names have no restrictions like identifiers do)
+            // so let us now concatenate all of the tokens from the macro expansion
+            // up to the '>'' in order to get the file name
+            SourceLocation rootExpansionLoc = sourceManager.getExpansionLoc(fileName.location());
+            Token concatenatedFileName = fileName;
+            std::vector<Token> tokens;
+            tokens.push_back(fileName);
```

Comment:
Prefer a SmallVectorSized<> here instead of an std::vector. in the common case where there aren't many entries, we avoid a heap allocation.